File: src/org/ice1000/devkt/ui/PsiViewer.java
package org.ice1000.devkt.ui;

import org.ice1000.devkt.psi.PsiViewerImpl;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;

/**
 * @author ice1000
 * @see PsiViewerImpl
 */
public abstract class PsiViewer extends JDialog {
	protected JPanel mainPanel;
	protected JButton buttonClose;
	protected JScrollPane pane;

	public PsiViewer(@Nullable Window owner) {
		super(owner);
	}
}

File: src/org/ice1000/devkt/psi/psi-viewer.kt
package org.ice1000.devkt.psi

import org.ice1000.devkt.ui.PsiViewer
import org.jetbrains.kotlin.com.intellij.psi.PsiElement
import org.jetbrains.kotlin.psi.KtFile
import java.awt.Window
import javax.swing.JTree
import javax.swing.tree.DefaultMutableTreeNode

typealias UINode = DefaultMutableTreeNode

/**
 * @author ice1000
 */
class PsiViewerImpl(file: KtFile, owner: Window? = null) : PsiViewer(owner) {
	init {
		contentPane = mainPanel
		isModal = true
		rootPane.defaultButton = buttonClose
		pane.setViewportView(JTree(mapAst2Display(file)))
		buttonClose.addActionListener { dispose() }
		pack()
	}

	/**
	 * 缅怀一下天国的 Lice AST Viewer
	 * TODO make the second arg of [cutText] configurable
	 */
	private fun mapAst2Display(
			node: PsiElement,
			root: UINode = UINode("${cutText(node.text, 30)} => $node")): UINode = when {
		node.firstChild == null -> UINode(node)
		else -> root.apply {
			node.children.forEach { add(mapAst2Display(it)) }
		}
	}
}
